### PR TITLE
Autoinstall fixes

### PR DESCRIPTION
**Pull request by anbe42**
This is how I expect `autoinstall` to work in case of modules being skipped due to `BUILD_EXCLUSIVE_*`.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -35,6 +35,7 @@ clean:
 
 dkms: dkms.in
 	sed -e 's/#RELEASE_STRING#/$(RELEASE_STRING)/' $^ > $@
+	chmod +x $@
```
Comment (evelikov) [inline Makefile:38]: The execute bit is already set by the install command further down. Plus last time I've looked using in-tree dkms was causing various issues.

File: dkms.in [hunk 2]
```diff
@@ -2261,7 +2261,7 @@ autoinstall() {
                     installed_modules[${#installed_modules[@]}]="$m"
                     install_count=$(($install_count +1))
                 else
-                    failed_modules=[${#failed_modules[@]}]="$m"
+                    failed_modules[${#failed_modules[@]}]="$m($?)"
```
Comment (evelikov) [inline dkms.in]: Nicely found. Out of curiosity, how did you spot it - code review or some tool/bash warning option?

Comment (anbe42) [reply dkms.in]: when I printed the array, it contained garbage ;-)

File: dkms.in [hunk 3]
```diff
@@ -1012,7 +1012,7 @@ prepare_build()
     read_conf_or_die "$kernelver" "$arch"
 
     # Error out if build_exclude is set
-    [[ $build_exclude ]] && die 9 \
+    [[ $build_exclude ]] && die 77 \
```
Comment (evelikov) [inline dkms.in:1018]: Is @xnox in the loop about this change? I believe he's a fellow maintainer in Debian/Ubuntu, correct?

I'm in favour of this change. Small request - let's have a small note in the man page. Thus anyone else relying on the old error code has a clear reference.

File: dkms.in [hunk 4]
```diff
@@ -2216,24 +2214,15 @@ autoinstall() {
...
-
-    # We may have exited the loop with $tenative set.  If it is,
-    # it contains something that should be updated.
-    if [[ $tenative ]]; then
-        to_install[${#to_install[@]}]="$tenative"
+        # Otherwise, autoinstall the latest version we have hanging around.
+        to_install[${#to_install[@]}]="$m/$v"
         build_depends["$m"]="${BUILD_DEPENDS[@]}"
-    fi
+    done
```
Comment (evelikov) [inline dkms.in:2229]: At a glance - this patch seems unrelated to the rest of the PR. Let's split it a separate PR?

**Comment by evelikov** [comment]
Huge thanks for the series - in general it looks good. There are a few of minor comments and I'm in favour of this landing with those addressed.

File: run_test.sh [hunk 5]
```diff
@@ -308,7 +309,7 @@ run_with_expected_output dkms build -k "${KERNEL_VER}" -m dkms_test -v 1.0 << EO
 
 Building module:
 Cleaning build area...
-make -j$(nproc) KERNELRELEASE=${KERNEL_VER} -C /lib/modules/${KERNEL_VER}/build M=/var/lib/dkms/dkms_test/1.0/build...
+make -j${nproc} KERNELRELEASE=${KERNEL_VER} -C /lib/modules/${KERNEL_VER}/build M=/var/lib/dkms/dkms_test/1.0/build...
```
[the remark this answers is not in the document]
Comment (anbe42) [reply run_test.sh:314]: as parallelism is not really beneficial for the tests, we could just run them under `parallel_jobs=1` and `s/\$\(nproc\)/1/`

File: Makefile (hunk 1 above)
Comment (anbe42) [reply Makefile:38]: IMO in-tree dkms should work fine, otherwise its some bug to be fixed.

File: run_test.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (anbe42) [reply run_test.sh]: replaced the patch with the parallel_jobs=1 approach

File: dkms.8.in [hunk 6]
```diff
@@ -657,6 +657,17 @@ will error out. (dkms autoinstall will not fail on this condition but skip the
 respective modules.) For example, if you set it as ="^2\.4.*", your module would
 not be built for 2.6 or later kernels.
 .TP
+.B BUILD_EXCLUSIVE_KERNEL_MIN=
+and
+.B BUILD_EXCLUSIVE_KERNEL_MAX=
+These optional directives allow one to specify the minimal and maximal kernel
+versions supported by the module. If one (or both) of these are defined, the
+module will not be built for kernels outside the specified version limits.
```
[the remark this answers is not in the document]
Comment (anbe42) [reply dkms.8.in:669]: I've updated the documentation in some commits to hopefully clarify that a bit. If any of the BUILD_EXCLUSIVE_* says not to build the module, it won't be built. 

Maybe there should be a separate chapter about BUILD_EXCLUSIVE_* in the manpage ...

[the remark this answers is not in the document]
Comment (anbe42) [reply dkms.8.in:669]: I've no good idea for writing tests for BUILD_EXCLUSIVE_* values ...

File: dkms.8.in [hunk 7]
```diff
@@ -657,6 +657,17 @@ will error out. (dkms autoinstall will not fail on this condition but skip the
...
+.B BUILD_EXCLUSIVE_KERNEL_MIN could be set to the version where some kernel
+feature was introduced that is required by the module, e.g. ="3.4".
+.B BUILD_EXCLUSIVE_KERNEL_MAX could be set to the version where the module
+was integrated into the kernel mainline, e.g. ="4.12".
```
[the remark this answers is not in the document]
Comment (anbe42) [reply dkms.8.in]: simplified

File: run_test.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (evelikov) [reply run_test.sh]: Haven't looked all the corner-cases but I believe you're correct.

Then again, such change should really be a separate PR. A cleaner option would be to `export MAKEFLAGS="-j..."` at the top, instead of repeatedly passing it to each make invocation.

File: Makefile (hunk 1 above)
Comment (evelikov) [reply Makefile:38]: I would love it if using dkms in-tree "just works". Although currently we `make install` before doing any tests, plus there were cases where it relied on particular files on the rootfs ... IIRC it was `/var/lib/dkms/version` or alike.

Fixing this would be amazing, although it's orthogonal to the rest of this PR so let's keep it separate. Off the top of my head we want to:
   - add the execute bit (as you've done above)
   - remove the `make install` from the CI
   - ensure dkms is not installed in the CI containers, and `/etc/dkms/` && `/var/lib/dkms/` are missing
   - fix any test fallouts
   
 Thanks in advance

File: dkms.8.in (hunk 6 above)
[the remark this answers is not in the document]
Comment (evelikov) [reply dkms.8.in:669]: The updated documentation looks much clearer. Thanks o/

For the tests, we can have a few negative tests:
 - `BUILD_EXCLUSIVE_KERNEL_MIN=-6.1` (same for max) - error out, version cannot be negative (accidental typo?)
 - `BUILD_EXCLUSIVE_KERNEL_MIN=6.1` and `BUILD_EXCLUSIVE_KERNEL_MAX=5.10` - error out, invalid (flipped?) boundaries
 - `BUILD_EXCLUSIVE_KERNEL="^\5\.10"`  and `BUILD_EXCLUSIVE_KERNEL_MIN=5.9` and `BUILD_EXCLUSIVE_KERNEL_MIN=5.11` - skip (assuming I'm reading the documentation correctly)

File: dkms.in (hunk 4 above)
Comment (evelikov) [reply dkms.in:2229]: Right, the `simplify` threw me off the scent for a moment.

File: run_test.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (evelikov) [reply run_test.sh]: Split and addressed in separate PR.

File: dkms_autoinstaller [hunk 8]
```diff
@@ -47,7 +48,7 @@ case "$1" in
 			kernel=$(uname -r)
 		fi
 		if [ -f /etc/dkms/no-autoinstall ]; then
-			log_daemon_msg "$prog: autoinstall for dkms modules has been disabled"
+			log_action_msg "$prog: autoinstall for dkms modules has been disabled"
 		else
 			log_daemon_msg "$prog: running auto installation service for kernel $kernel"
```
[the remark this answers is not in the document]
Comment (evelikov) [reply dkms_autoinstaller:53]: Split and addressed in separate PR.

File: Makefile (hunk 1 above)
Comment (evelikov) [reply Makefile:38]: Split and addressed in separate PR. Essentially the `pwd` manipulation was removed.

File: dkms.in [hunk 9]
```diff
@@ -981,6 +981,7 @@ prepare_signing()
     # scripts/sign-file.c in kernel source also supports using "pkcs11:..." as private key
     if [[ "$mok_signing_key" != "pkcs11:"* ]] && ( [ ! -f "$mok_signing_key" ] || [ ! -f "$mok_certificate" ] ); then
         echo "Certificate or key are missing, generating self signed certificate for MOK..."
+        command -v openssl >/dev/null || echo "openssl not found, can't generate key and certificate."
```
Comment (evelikov) [inline dkms.in:984]: Nit: Please also add a "return" in this case.

Can you rebase the series against master and look at the CI failures. Having the `BUILD_EXCLUSIVE` combinations checked (as mentioned earlier) would be great, but can be done as follow-up.

Thanks again o/

**Comment by anbe42** [comment]
after eliminating a lot of nondeterminism in the output, I finally got it all green ;-)

File: run_test.sh [hunk 10]
```diff
@@ -215,19 +217,17 @@ run_with_expected_error() {
     local dkms_command="$3"
     local output_log=test_cmd_output.log
     local expected_output_log=test_cmd_expected_output.log
-    local error_code
+    local error_code=0
```
Comment (evelikov) [inline run_test.sh:233]: Unless I'm missing something `run_with_expected_output()` can now become a simple wrapper around `run_with_expected_error()`, correct?

If so, feel free to fix that in a future PR.

File: run_test.sh [hunk 11]
```diff
@@ -1319,7 +1319,7 @@ echo 'Running autoinstall with failing test module and test module with dependen
 run_with_expected_error 11 dkms autoinstall -k "${KERNEL_VER}" << EOF
 
 Building module:
-Cleaning build area...(bad exit status: 2)
+Cleaning build area...
```
Comment (evelikov) [inline run_test.sh:1337]: The (2x) bad exit status should squashed with another patch, right?

File: dkms.in [hunk 12]
```diff
@@ -2206,14 +2207,17 @@ autoinstall() {
...
+        elif [[ ("$(VER "$v")" > "$(VER "${latest["$m"]}")") ]]; then
             latest["$m"]="$v"
         fi
     done < <(module_status)
 
     # Walk through our list of known modules, and create
     # a list of modules that need to be reinstalled.
-    for m in "${!latest[@]}"; do
+    for m in "${known_modules[@]}"; do
         v="${latest["$m"]}"
```
Comment (evelikov) [inline dkms.in:2222]: It's no immediately obvious why the order matters at this point. All we do is gather a list of to install modules and their respective deps.

Can you add some context in the commit message?

File: run_test.sh (hunk 11 above)
Comment (anbe42) [reply run_test.sh:1337]: No. This patch fixes malformed Makefiles indented with spaces instead of tabs.

File: run_test.sh (hunk 10 above)
Comment (anbe42) [reply run_test.sh:233]: need to dig into that again at some point

File: run_test.sh [hunk 13]
```diff
@@ -191,12 +193,16 @@ genericize_expected_output() {
...
+    local stdout_log=test_cmd_stdout.log
+    local stderr_log=test_cmd_stderr.log
     local output_log=test_cmd_output.log
     local expected_output_log=test_cmd_expected_output.log
     local error_code=0
 
     cat > ${expected_output_log}
-    "$@" > ${output_log} 2>&1 || error_code=$?
+    "$@" > ${stdout_log} 2> ${stderr_log} || error_code=$?
+    cat ${stdout_log} ${stderr_log} > ${output_log}
```
[the remark this answers is not in the document]
Comment (anbe42) [reply run_test.sh:208]: Can we try that later?

File: dkms.in (hunk 12 above)
Comment (anbe42) [reply dkms.in:2222]: mainly for deterministic tests. if autoinstall is going to build more than one module (successful or not), order does matter for the output

File: run_test.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (evelikov) [reply run_test.sh:208]: Sure thing, I'm just going to flip this back to unresolved for posterity.

[the remark this answers is not in the document]
Comment (anbe42) [reply run_test.sh:208]: I managed to deterministically interleave stdout with stderr using `stdbuf`, will be prepended to the next PR.

File: run_test.sh (hunk 10 above)
Comment (anbe42) [reply run_test.sh:233]: `run_with_expected_output` can just call `run_with_expected_error 0 "$@"`, prepended to next PR.
